### PR TITLE
[Doc] Add data structure table by datamodel

### PR DESCRIPTION
**Pull request by mapedraza**
This PR adds a section with a summary of the database data structure for each datamodel available

### CODE REVIEW

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md [hunk 1]
```diff
@@ -116,6 +116,22 @@ Please observe the concatenation of entity ID and type is already given in the `
...
+|:----|:----|:----|:----|
+|dm-by-fixed-entity-type-database-schema|fiware-service|fiware-servicePath|entity type|
+|dm-by-fixed-entity-type-database|fiware-service|fiware-service|entity type|
+|dm-by-fixed-entity-type|agent properties|fiware-service|entity type|
+|dm-by-entity-database-schema|fiware-service|fiware-servicePath|fiware-servicePath + entity type + entity id|
+|dm-by-entity-database|fiware-service|fiware-service|fiware-servicePath + entity type + entity id|
+|dm-by-entity-type|agent properties|fiware-service|fiware-servicePath + entity type|
+|dm-by-service-path|agent properties|fiware-service|fiware-servicePath|
+|dm-by-entity|agent properties|fiware-service|fiware-servicePath + entity type + entity id|
+
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:133]:
```suggestion

In the above table `+` means concatenation using the `_` character.
```

Moreover, I think "agent properties" should be explained a bit more. I guess it comes from "some setting in agent.conf Cygnus configuration file" but it should be detailed a little bit more.

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md [hunk 2]
```diff
@@ -116,6 +116,22 @@ Please observe the concatenation of entity ID and type is already given in the `
...
+This table contains the database structure used for each datamodel supported by this sink
+
+|data_model|database name|schema name|table name|
+|:----|:----|:----|:----|
+|dm-by-fixed-entity-type-database-schema|fiware-service|fiware-servicePath|entity type|
+|dm-by-fixed-entity-type-database|fiware-service|fiware-service|entity type|
+|dm-by-fixed-entity-type|agent properties|fiware-service|entity type|
+|dm-by-entity-database-schema|fiware-service|fiware-servicePath|fiware-servicePath + entity type + entity id|
+|dm-by-entity-database|fiware-service|fiware-service|fiware-servicePath + entity type + entity id|
+|dm-by-entity-type|agent properties|fiware-service|fiware-servicePath + entity type|
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:130]:
```suggestion
|dm-by-entity-type-database-schema|fiware-service|fiware-servicePath|fiware-servicePath + entity type|
|dm-by-entity-type-database|fiware-service|fiware-service|fiware-servicePath + entity type|
|dm-by-entity-type|agent properties|fiware-service|fiware-servicePath + entity type|
```

**Comment by mapedraza** [comment]
After some tests, the behaviour checked is the following:

| datamodel|database|schema|table|note|
| ---------|--------|------|-----|----|
| by default| agent.conf | service | servicepath_entityname_entitytype | Automatic schema creation |
| dm-by-fixed-entity-type-database-schema| service | _servicepath | entitytype | Database and schema need to be created in advance
| dm-by-fixed-entity-type-database	| service | service | entitytype | Database and schema need to be created in advance
| dm-by-fixed-entity-type	| agent.conf | service | entitytype | Automatic schema creation |
| dm-by-entity-database-schema	| service | _servicepath  | servicepath_entityname_entitytype | Database and schema need to be created in advance |
| dm-by-entity-database	| service | service | servicepath_entityname_entitytype | Database and schema need to be created in advance|
| dm-by-entity-type	| agent.conf | service | servicepath_entitytype | Automatic schema creation |
| dm-by-service-path	| agent.conf | service | servicepath | Automatic schema creation| 
| dm-by-entity	| agent.conf | service | servicepath_entityname_entitytype | Automatic schema creation |

where agent.conf means the value of the configuration paramenter `cygnus-ngsi.sinks.postgresql-sink.postgresql_database` or value of the environment var `CYGNUS_POSTGRESQL_DATABASE`

**Comment by fgalan** [comment]
> After some tests, the behaviour checked is the following:
> datamodel 	database 	schema 	table 	note
> by default 	agent.conf 	service 	servicepath_entityname_entitytype 	Automatic schema creation
> dm-by-fixed-entity-type-database-schema 	service 	_servicepath 	entitytype 	Database and schema need to be created in advance
> dm-by-fixed-entity-type-database 	service 	service 	entitytype 	Database and schema need to be created in advance
> dm-by-fixed-entity-type 	agent.conf 	service 	entitytype 	Automatic schema creation
> dm-by-entity-database-schema 	service 	_servicepath 	servicepath_entityname_entitytype 	Database and schema need to be created in advance
> dm-by-entity-database 	service 	service 	servicepath_entityname_entitytype 	Database and schema need to be created in advance
> dm-by-entity-type 	agent.conf 	service 	servicepath_entitytype 	Automatic schema creation
> dm-by-service-path 	agent.conf 	service 	servicepath 	Automatic schema creation
> dm-by-entity 	agent.conf 	service 	servicepath_entityname_entitytype 	Automatic schema creation

Is the same as originally documented? Or there are differences?

> where agent.conf means the value of the configuration paramenter `cygnus-ngsi.sinks.postgresql-sink.postgresql_database` or value of the environment var `CYGNUS_POSTGRESQL_DATABASE`

Ok. I'd suggest to tell it explicitly (maybe with a bottom note).

**Comment by mapedraza** [comment]
> > After some tests, the behaviour checked is the following:
> > datamodel 	database 	schema 	table 	note
> > by default 	agent.conf 	service 	servicepath_entityname_entitytype 	Automatic schema creation
> > dm-by-fixed-entity-type-database-schema 	service 	_servicepath 	entitytype 	Database and schema need to be created in advance
> > dm-by-fixed-entity-type-database 	service 	service 	entitytype 	Database and schema need to be created in advance
> > dm-by-fixed-entity-type 	agent.conf 	service 	entitytype 	Automatic schema creation
> > dm-by-entity-database-schema 	service 	_servicepath 	servicepath_entityname_entitytype 	Database and schema need to be created in advance
> > dm-by-entity-database 	service 	service 	servicepath_entityname_entitytype 	Database and schema need to be created in advance
> > dm-by-entity-type 	agent.conf 	service 	servicepath_entitytype 	Automatic schema creation
> > dm-by-service-path 	agent.conf 	service 	servicepath 	Automatic schema creation
> > dm-by-entity 	agent.conf 	service 	servicepath_entityname_entitytype 	Automatic schema creation
> 
> Is the same as originally documented? Or there are differences?

Overall the same, just differences in table order naming ( servicepath_entityname_entitytype instead of servicepath_entitytype_entityname) or schema name when using service-path (_servicepath)

> > where agent.conf means the value of the configuration paramenter `cygnus-ngsi.sinks.postgresql-sink.postgresql_database` or value of the environment var `CYGNUS_POSTGRESQL_DATABASE`
> 
> Ok. I'd suggest to tell it explicitly (maybe with a bottom note).

Ok

**Comment by mapedraza** [comment]
I also suggest to include the column notes with the automatic schema creation

**Comment by fgalan** [comment]
> I also suggest to include the column notes with the automatic schema creation

Yes. The "Notes" column is a good improvement :)

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md (hunk 1 above)
Comment (mapedraza) [reply doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md]: Fixed in https://github.com/telefonicaid/fiware-cygnus/pull/2196/commits/df6b3242b05629932240c1345b6a11ee0ff6d884

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md (hunk 2 above)
Comment (mapedraza) [reply doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md]: Added in https://github.com/telefonicaid/fiware-cygnus/pull/2196/commits/df6b3242b05629932240c1345b6a11ee0ff6d884

**Comment by fgalan** [comment]
It would be great to review documentation and remove duplicated description, now covered by the curated table.

**Comment by mapedraza** [comment]
> It would be great to review documentation and remove duplicated description, now covered by the curated table.

Done in https://github.com/telefonicaid/fiware-cygnus/pull/2196/commits/810e409bd100a71ac1d84527c4ca123d4124e0c4 and https://github.com/telefonicaid/fiware-cygnus/pull/2196/commits/d8bb272ff8724791fe6bbbeb261b46be97b7cffa

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md [hunk 3]
```diff
@@ -45,58 +45,50 @@ This is done at the cygnus-ngsi Http listeners (in Flume jergon, sources) thanks
...
+| `dm-by-fixed-entity-type-database`        | service       | service      | entitytype                        | Database and schema need to be created in advance |
+| `dm-by-fixed-entity-type`                 | `agent.conf`  | service      | entitytype                        | Automatic schema creation                         |
+| `dm-by-entity-database-schema`            | service       | _servicepath | servicepath_entityname_entitytype | Database and schema need to be created in advance |
+| `dm-by-entity-database`                   | service       | service      | servicepath_entityname_entitytype | Database and schema need to be created in advance |
+| `dm-by-entity-type`                       | `agent.conf`  | service      | servicepath_entitytype            | Automatic schema creation                         |
+| `dm-by-service-path`                      | `agent.conf`  | service      | servicepath                       | Automatic schema creation                         |
+| `dm-by-entity`                            | `agent.conf`  | service      | servicepath_entityname_entitytype | Automatic schema creation                         |
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:65]: This table has 8 entries. However, it seems Cygnus uses 11 dm-by-xxx literals

https://github.com/telefonicaid/fiware-cygnus/blob/aae3eb8ca471f2c3b5aca424c8c9175983aa7fc9/cygnus-common/src/main/java/com/telefonica/iot/cygnus/utils/CommonUtils.java#L237-L262

So it seems 3 are missing in this table

File: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md [hunk 4]
```diff
@@ -45,58 +45,53 @@ This is done at the cygnus-ngsi Http listeners (in Flume jergon, sources) thanks
...
+
+For a given entity, `entityname`, of type `entitytype`, inside a tenant `service` under `/servicepath`, how the information is stored in the database differs from using different datamodel configurations. The following table contains the database, schema and table names used for each datamodel supported by this sink:
+
+| datamodel name                            | database name | schema  name | table name                        | note                                              |
+|-------------------------------------------|---------------|--------------|-----------------------------------|---------------------------------------------------|
+| by default                                | `agent.conf`  | service      | servicepath_entityname_entitytype | Automatic schema creation                         |
+| `dm-by-fixed-entity-type-database-schema` | service       | _servicepath | entitytype                        | Database and schema need to be created in advance |
```
Comment (fgalan) [inline doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:58]: The initial underscore `_` in `-schema` DM is a bit weird (although consistent with the code implementation)

Checking with @manucarrace before merging this PR.

Comment (fgalan) [reply doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_postgresql_sink.md:58]: Discussed with @manucarrace by chat

He was unaware of this. The `-schema` DMs are not used.

An issue has been created to analyze this: https://github.com/telefonicaid/fiware-cygnus/issues/2201

**Review by fgalan — APPROVED** [review]
LGTM. Impressive work! :)

**Pull request merged**